pandas/core/indexes/base.py

```diff
@@ -259,12 +261,22 @@ def __new__(
         dtype=None,
         copy=False,
         name=None,
+        names=None,
         fastpath=None,
         tupleize_cols=True,
         **kwargs
     ):
 
-        if name is None and hasattr(data, "name"):
+        if names is not None:
```

Comment:
Not super familiar with this part of the code but is it possible to share the length of checking of `names` across the index types? This already fails for a MultiIndex:

```python
>>> pd.MultiIndex.from_tuples((('a', 'b'), ('c', 'd')), names=('a', 'b', 'c'))

ValueError: Length of names must match number of levels in MultiIndex.
```

So ideally could share that machinery here